src/mpid/ch4/netmod/ucx/ucx_am.h

```diff
@@ -258,25 +305,52 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_NM_am_isend_reply(MPIR_Context_id_t context_i
     ucx_hdr.handler_id = handler_id;
     ucx_hdr.data_sz = data_sz;
 
-    /* just pack and send for now */
-    send_buf = MPL_malloc(data_sz + am_hdr_sz + sizeof(ucx_hdr), MPL_MEM_BUFFER);
-    MPIR_Memcpy(send_buf, &ucx_hdr, sizeof(ucx_hdr));
-    MPIR_Memcpy(send_buf + sizeof(ucx_hdr), am_hdr, am_hdr_sz);
     if (dt_contig) {
-        MPIR_Memcpy(send_buf + am_hdr_sz + sizeof(ucx_hdr), (char *) data + dt_true_lb, data_sz);
+        if (data_sz >= MPIR_CVAR_CH4_UCX_AM_IOV_THRESH) {
+            send_buf = MPL_malloc(sizeof(ucx_hdr) + am_hdr_sz, MPL_MEM_BUFFER);
+            MPIR_Memcpy(send_buf, &ucx_hdr, sizeof(ucx_hdr));
+            MPIR_Memcpy(send_buf + sizeof(ucx_hdr), am_hdr, am_hdr_sz);
+
+            iov[0].buffer = (void *) send_buf;
+            iov[0].length = sizeof(ucx_hdr) + am_hdr_sz;
+            iov[1].buffer = (void *) ((char *) data + dt_true_lb);
+            iov[1].length = data_sz;
+
+            dt = ucp_dt_make_iov();
+            tot_sz = 2;
+        } else {
+            /* just pack and send for now */
+            send_buf = MPL_malloc(data_sz + am_hdr_sz + sizeof(ucx_hdr), MPL_MEM_BUFFER);
+            MPIR_Memcpy(send_buf, &ucx_hdr, sizeof(ucx_hdr));
+            MPIR_Memcpy(send_buf + sizeof(ucx_hdr), am_hdr, am_hdr_sz);
+            MPIR_Memcpy(send_buf + am_hdr_sz + sizeof(ucx_hdr), (char *) data + dt_true_lb,
+                        data_sz);
+
+            dt = ucp_dt_make_contig(1);
+            tot_sz = data_sz + am_hdr_sz + sizeof(ucx_hdr);
+        }
```

Comment:
Currently the threshold for IOV is equal to rendezvous threshold so this could be replaced with IOV only

Reply:
Are we sure that UCX doesn't already do threshold checking and packing for small/medium messages?

Reply:
I am not sure I understand the comment. What UCX does underneath should be independent from the UCX netmod isn't it? There are several protocols that UCX itself uses but all these work with whatever buffer the UCX netmod passes down to it (either packed or unpacked). I know that for small messages (between `128` and `1024` bytes) it uses `ucp_eager_bcopy_single` protocol in which UCX netmod's buffer is copied to a preregistered buffer (allocated at startup). More information on different protocols is contained in the presentation above.

Reply:
I'm re-reading the comments and trying to swap back in all the context for these changes. I guess I'm trying to understand why our packing is better than UCX's packing. Why is there so much less overhead to ours?

Reply:
Although we already discussed this in person I am going to write it down for future references. The reason IOVEC performs worse than UCX netmod packing into `send_buf` is because in this case when using the `ucp_dt_make_contig(1)` datatype UCX can use `ucp_tag_eager_contig_short`. This protocol does message inlining for sizes between 0 and 64 bytes. When using IOVEC (i.e., `ucp_dt_make_iov()`) the protocol used by UCX is `ucp_eager_bcopy_single` which copies user data (as well as UCX AM header) into a preregister buffer picked from a memory pool. This translates in a performance penalty of ~30% in latency for the IOVEC vs. UCX netmod packing.